libamqpprox/amqpprox_connectionlimitermanager.cpp

```diff
@@ -0,0 +1,266 @@
+/*
+** Copyright 2022 Bloomberg Finance L.P.
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+
+#include <amqpprox_connectionlimitermanager.h>
+
+#include <amqpprox_connectionlimiterinterface.h>
+#include <amqpprox_fixedwindowconnectionratelimiter.h>
+#include <amqpprox_logging.h>
+
+#include <memory>
+#include <string>
+
+namespace Bloomberg {
+namespace amqpprox {
+
+ConnectionLimiterManager::ConnectionLimiterManager()
+: d_connectionRateLimitersPerVhost()
+, d_alarmOnlyConnectionRateLimitersPerVhost()
+, d_defaultConnectionRateLimit(0)
+, d_defaultAlarmOnlyConnectionRateLimit(0)
+, d_mutex()
+{
+}
+
+std::shared_ptr<ConnectionLimiterInterface>
+ConnectionLimiterManager::addConnectionRateLimiter(
+    const std::string &vhostName,
+    uint32_t           numberOfConnections)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    std::shared_ptr<FixedWindowConnectionRateLimiter> connectionRateLimiter =
+        std::make_shared<FixedWindowConnectionRateLimiter>(
+            numberOfConnections);
+    d_connectionRateLimitersPerVhost[vhostName] = {true,
+                                                   connectionRateLimiter};
+    return connectionRateLimiter;
+}
+
+std::shared_ptr<ConnectionLimiterInterface>
+ConnectionLimiterManager::addAlarmOnlyConnectionRateLimiter(
+    const std::string &vhostName,
+    uint32_t           numberOfConnections)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    std::shared_ptr<FixedWindowConnectionRateLimiter>
+        alarmOnlyConnectionRateLimiter =
+            std::make_shared<FixedWindowConnectionRateLimiter>(
+                numberOfConnections);
+    d_alarmOnlyConnectionRateLimitersPerVhost[vhostName] = {
+        true, alarmOnlyConnectionRateLimiter};
+    return alarmOnlyConnectionRateLimiter;
+}
+
+void ConnectionLimiterManager::setDefaultConnectionRateLimit(
+    uint32_t defaultConnectionRateLimit)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    d_defaultConnectionRateLimit = defaultConnectionRateLimit;
+    for (auto &limiter : d_connectionRateLimitersPerVhost) {
+        if (!limiter.second.first) {
```

Comment:
would prefer to see something that searches and makes use of the resulting iterator set e.g. `equal_range` or similar rather than iterating over the full set of items in the container.

Reply:
I think the `equal_range` function will only be applied when we want to search based on the key. Here we are iterating to find the entry which has default rate limiter entries. So basically, we are searching for specific value for all the vhost keys. And the size of the map depends upon the number of vhosts. And this particular function, we are only going to call from control command. We can have separate data member to maintain the default limiters for each vhost. But that will also increase overhead to maintain the new data structure.